scheduling/allocator.rb

```diff
@@ -315,56 +375,106 @@ def update(vm, vm_host)
   # The VmHostSliceAllocation is used when system is configured
   # to allocate VMs inside VmHostSlice. It wraps around the VmHostAllocation class
   # used otherwise.
-  # When creating a dedicated slice for a VM, as in the Standard family, we look
-  # for a place for a new slice and select a cpu set that will be used for that.
+  # This class handles two distinct cases - one when a dedicated slice
+  # is created for a VM (as in Standard family). In that case it always looks
+  # for a place for a new slice and finds a cpu set that will be used for that.
+  # The other case is when a shared slice can be used, meaning multiple VMs can be
+  # placed inside a slice. In that case the Allocation first looks if the candidate host
+  # already has a slice allocated that matches the family of the VM and that
+  # has enough room. If so, it will use that. Otherwise, a new slice will be
+  # created same way as in the 'dedicated' case.
+  # All this logic is split between 'select_existing_slice' and 'update' methods. The first
+  # one checks if the candidate_host can host a slice or has one already.
+  # The second one creates a slice if needed, once the host candidate is selected.
   class VmHostSliceAllocation
     def initialize(candidate_host, request, vm_host_allocations)
       @candidate_host = candidate_host
       @request = request
       @vm_host_allocations = vm_host_allocations
+
+      @existing_slice = if @request.require_shared_slice then select_existing_slice end
     end
 
     def is_valid
-      @vm_host_allocations.all?(&:is_valid)
+      if @existing_slice
+        true
+      else
+        @vm_host_allocations.all?(&:is_valid)
+      end
     end
 
     def utilization
-      # compute the score based on combined CPU and Memory utilization, as usual
-      util = @vm_host_allocations.map(&:utilization)
-      util.sum.fdiv(util.size)
+      if @existing_slice
+        # if we found an existing slice, return the desired utilization
+        # to make this a preferred choice
+        @request.target_host_utilization
+      else
+        # otherwise, compute the score based on combined CPU and Memory utilization, as usual
+        util = @vm_host_allocations.map(&:utilization)
+        util.sum.fdiv(util.size)
+      end
     end
 
     def update(vm, vm_host)
-      DB.transaction do
-        cpus = select_cpuset(vm_host.id, @request.vcpus)
-
-        st = Prog::Vm::VmHostSliceNexus.assemble_with_host(
-          "#{vm.family}_#{vm.inhost_name}",
-          vm_host,
-          family: vm.family,
-          allowed_cpus: cpus,
-          memory_gib: @request.memory_gib_for_vcpus(cpus.count),
-          is_shared: false
-        )
+      if @existing_slice.nil?
+        DB.transaction do
```

Comment:
Probably best to move this around the preceding `if`, so you only need one call and not a separate call in both the `if` and `else` branches.